cmd/adapter/config.go

```diff
@@ -0,0 +1,272 @@
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/logrange/logrange/pkg/forwarder"
+	"github.com/logrange/logrange/pkg/utils"
+	"github.com/logrange/range/pkg/transport"
+	"io/ioutil"
+)
+
+type (
+	gravity struct {
+		ApiListenAddr string
+		Kubernetes    *kube
+	}
+
+	kube struct {
+		Namespace              string
+		ForwarderConfigMapName string
+	}
+
+	// cronQuery describe a query, which should be called periodically
+	cronQuery struct {
+		// Query contains LQL which should be called periodically
+		Query string
+		// PeriodSec defines an interval in seconds to execute the Query
+		PeriodSec int
+	}
+
+	logrange struct {
+		CronQueries       []cronQuery
+		Partition         string
+		ForwarderTmplFile string
+		Kubernetes        *kube
+		Transport         *transport.Config
+	}
+
+	Config struct {
+		Gravity         *gravity
+		Logrange        *logrange
+		SyncIntervalSec int
+	}
+)
+
+func LoadCfgFromFile(path string) (*Config, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	cfg := &Config{}
+	err = json.Unmarshal(data, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+//===================== config =====================
+
+func NewDefaultConfig() *Config {
+	return &Config{
+		Gravity: &gravity{
+			ApiListenAddr: "127.0.0.1:8083",
+			Kubernetes: &kube{
+				Namespace:              "kube-system",
+				ForwarderConfigMapName: "log-forwarders",
+			},
+		},
+
+		Logrange: &logrange{
+			Partition:         "logrange.pipe=__default__",
+			ForwarderTmplFile: "/opt/logrange/gravity/config/forward-tmpl.json",
+			Kubernetes: &kube{
+				Namespace:              "kube-system",
+				ForwarderConfigMapName: "lr-forwarder",
+			},
+			Transport: &transport.Config{
+				ListenAddr: "127.0.0.1:9966",
+			},
+		},
+
+		SyncIntervalSec: 5,
+	}
+}
+
+func (c *Config) Apply(other *Config) {
+	if other == nil {
+		return
+	}
+
+	if other.Gravity != nil {
+		c.Gravity.apply(other.Gravity)
+	}
+	if other.Logrange != nil {
+		c.Logrange.apply(other.Logrange)
+	}
+	if other.SyncIntervalSec != 0 {
+		c.SyncIntervalSec = other.SyncIntervalSec
+	}
+}
+
+func (c *Config) check() error {
+	if c.Gravity == nil {
+		return fmt.Errorf("invalid Gravity=%v, must be non-nil", c.Gravity)
+	}
+	if c.Logrange == nil {
+		return fmt.Errorf("invalid Logrange=%v, must be non-nil", c.Logrange)
+	}
+	if err := c.Gravity.check(); err != nil {
+		return fmt.Errorf("invalid Gravity=%v: %v", c.Gravity, err)
+	}
+	if err := c.Logrange.check(); err != nil {
+		return fmt.Errorf("invalid Logrange=%v: %v", c.Logrange, err)
+	}
+	if c.SyncIntervalSec <= 0 {
+		return fmt.Errorf("invalid SyncIntervalSec=%v, must be > 0sec", c.SyncIntervalSec)
+	}
+	return nil
+}
+
+func (c *Config) String() string {
+	return utils.ToJsonStr(c)
+}
+
+//===================== gravity =====================
+
+func (g *gravity) apply(other *gravity) {
```

Comment:
`apply` seems like a substitute to `merge` (more justified name imho, due to exhibited behavior).

Reply:
renamed, applied (not yet merged to this branch)